### PR TITLE
[UIE-51] Move FocusTrap to components

### PR DESCRIPTION
**Pull request by nawatts**
https://broadworkbench.atlassian.net/browse/UIE-51

Continuing to move basic components to the components package... This moves FocusTrap.

I couldn't resist a little renaming. Changed "FocusTrapper" to "FocusTrap". "\<noun\>" vs "\<verb\> doer" fits better with React's declarative model.

Also changed "onBreakout" to "onEscape". I think that preserves the meaning (escaping the trap vs breaking out of the trap) and also makes it clearer how it's triggered (by pressing the escape key).

Also added a test.

### CODE REVIEW

File: src/components/common/FocusTrapper.ts [hunk 1]
```diff
@@ -1,36 +0,0 @@
-import _ from 'lodash/fp';
-import FocusLock from 'react-focus-lock';
-import { h } from 'react-hyperscript-helpers';
-
-// react-focus-lock does not export a type for FocusLock's props, but since FocusLock
-// is a function component, we can get the type from its parameters.
-type FocusLockProps = Parameters<typeof FocusLock>[0];
-
-export interface FocusTrapperProps extends FocusLockProps {
```
Comment (nawatts) [inline src/components/common/FocusTrapper.ts:9]: This was wrong. FocusTrapper doesn't pass on props to the underlying FocusLock, it passes them to its `lockProps`, where they're passed to the rendered div element.

File: packages/components/src/FocusTrap.test.ts [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+import { button, h } from 'react-hyperscript-helpers';
+
+import { FocusTrap } from './FocusTrap';
+
+describe('FocusTrap', () => {
+  it('traps focus', async () => {
+    // Arrange
+    const user = userEvent.setup();
+
+    const { container } = render(h(FocusTrap, [button(['A']), button(['B']), button(['C'])]));
```
Comment (msilva-broad) [inline packages/components/src/FocusTrap.test.ts]: really nice test, but how about we add a Button 'D' outside the focus-trap to really validate that the trap is functioning.

Comment (nawatts) [reply packages/components/src/FocusTrap.test.ts]: Excellent idea. Done in https://github.com/DataBiosphere/terra-ui/pull/4282/commits/7aed1dfef0c5d47f8b08446f605230c9a244c2eb.

**Comment by nawatts** [comment]
Looks like there's an issue that's causing Terra UI tests to fail:

The CommonJS build needs to
```ts
const FocusLock = require("react-focus-lock").default;
```
instead of
```
const FocusLock = require("react-focus-lock");
```

Marking this as a draft while I figure out how to work around that.

File: packages/build-utils/src/vite.config.mjs [hunk 3]
```diff
@@ -29,6 +29,9 @@ export default defineConfig(({ mode }) => {
         // Do not bundle dependencies.
         external: /node_modules/,
         output: {
+          // Handle default exports from CommonJS dependencies.
+          // https://rollupjs.org/configuration-options/#output-interop
+          interop: 'auto',
```
Comment (nawatts) [inline packages/build-utils/src/vite.config.mjs:34]: Without this, the CommonJS output contains:
```js
const FocusLock = require("react-focus-lock");
```

when what we want is something like:
```js
const FocusLock = require("react-focus-lock").default;
```

File: packages/components/src/FocusTrap.test.ts [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    // Focus should go through the buttons, then to the wrapper element, and then back to button A, never reaching button D
+    wrapper.focus();
+    const tabOrder: Element[] = [];
+    for (let i = 0; i < 5; i += 1) {
+      await user.tab();
+      tabOrder.push(document.activeElement!);
+    }
+
+    // Assert
+    expect(tabOrder).toEqual([buttonA, buttonB, buttonC, wrapper, buttonA]);
```
Comment (dmohs) [inline packages/components/src/FocusTrap.test.ts:31]: I love this! Testing accessibility!

File: src/components/PopupTrigger.js [hunk 5]
```diff
@@ -98,7 +98,7 @@ const PopupTrigger = forwardRefWithName(
             },
             ...props,
           },
-          [h(FocusTrapper, { onBreakout: () => setOpen(false) }, [content])]
+          [h(FocusTrap, { onEscape: () => setOpen(false) }, [content])]
```
Comment (dmohs) [inline src/components/PopupTrigger.js:101]: Agree with the naming changes.

**Review by dmohs — APPROVED** [review]

**Review by msilva-broad — APPROVED** [review]
really nice work!

**Pull request merged**